Add SecureDrop installation test

deeplow opened the pull request:
First attempt at adding a test for SecureDrop.

Commit:
Add SecureDrop installation test

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,51 @@
+    assert_and_click("menu-vm-xterm");
+
+
+    assert_script_run('gpg --keyserver hkps://keys.openpgp.org --recv-key "2359 E653 8C06 13E6 5295 5E6C 188E DD3B 7B22 E6A3"');
```
`assert_script_run` depends on seeing serial console output - serial console from "work" VM isn't directly connected to the one of the host; for this to work you either need to run something like `tail -F /var/log/xen/console/guest-work.log >> /dev/hvc0` in dom0 (we do that [here](https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/guivm_startup.pm#L43)), or do all that from dom0's terminal via qvm-run

marmarek replied:
Yes, that would work, but your test wouldn't detect if any of those command fails (other than possible some later step dom0 in dom0 failing).

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,51 @@
+        click_lastmatch();
+    }
+    assert_and_click("menu-vm-work");
+    assert_and_click("menu-vm-xterm");
```
and it failed: https://openqa.qubes-os.org/tests/111752#step/install/6

You hit a trap :P `xterm` isn't in the menu by default, it's added in https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/simple_gui_apps.pm. In tests that use xterm opened in a VM, I usually add a comment about it: https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/screenlocker_idle_watch.pm#L22

marmarek replied:
Xfce4-terminal should be fine. You'll need needle for the menu and then for the actual window (see any of the tests using xterm)

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,54 @@
+    assert_script_run('curl https://raw.githubusercontent.com/freedomofpress/securedrop/d91dc67/securedrop/tests/files/test_journalist_key.sec.no_passphrase | sudo tee /usr/share/securedrop-workstation-dom0-config/sd-journalist.sec');
+    assert_script_run('sdw-admin --validate');
+
+    assert_script_run('xfce4-power-manager -q'); # disable screen blanking during long command
```
My notes have this line:
```
x11_start_program('env xset s off', valid => 0);
```
but I'm not sure if that was enough either.

marmarek replied:
It unblanked on the key press.

marmarek commented:
Anyway:
```
# Test died: command 'sdw-admin --apply' timed out at /usr/lib/os-autoinst/autotest.pm line 412.
```
So, longer timeout? This is running virtualized, so runs slower than native.

And also, I recommend collecting and uploading logs. For example wrap it with `script`, or use `tee` (see https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/update2.pm#L110-L111 for example). You can also do post-fail hook to collect extra info on failure too: https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/update2.pm#L168-L174

marmarek commented:
Have you restarted the test after adding the needle? Or did you added it via developer mode?

marmarek commented:
I see the issue: you haven't added the `securedrop-launcher` tag, it only has `desktop` tag (which shouldn't be there I think). I guess you added it by clicking on an earlier screenshot (you can do that too, but then you need to adjust tags manually, as the default will be about that other screenshot).

marmarek commented:
For this one I just edited it manually.
But generally create new one, and don't be afraid about adding tags. In fact, do add more of them :) for SD-specific needles add `ENV-securedrop` tag.

marmarek commented:
> @marmarek after this test is finished, I think it would be nice if we could preserve the last version of this installation to serve as a base for other tests. Is that something you're OK with us doing?

Yes, absolutely.

> From what I could find, it looks like we need this job to have `PUBLISH_HDD_1` and then the other ones to use this as `HDD_1`. This is probably already what's happening on some builds. But I'm guessing that's on the admin interface and no so much something I can do within the test.

Technically, I think you can do the necessary change via the API, but sure, let me configure it.

> Although on my part I probably need to change `main.pm` to have the SecureDrop test run before the `STORE_HDD_1`.

Yes, definitely.

And few hints from my experience:
1. Upload the install log after successful install too - one day it will be useful to inspect what actually got tested.
2. In similar spirit, add collecting packages versions - copy&paste this snippet: https://github.com/QubesOS/openqa-tests-qubesos/blob/7242736b1ae34d9de31e906618ff48055bc853e5/tests/update2.pm#L121-L138 (need to add also `use Mojo::File qw(path);`)

> On this note, it would be nice if OpenQA also captured the RAM hybernation-style, but I didn't find any docs on this. So I'm guessing it's not a thing.

There is support for snapshots, but those are only to rollback to known-good state after one test module fail, to continue with further tests without restarting everything.  When I tried it some (long) time ago, it was rather unreliable with qubes, so we have this feature disabled. There is no support for preserving RAM between different test jobs.

marmarek commented:
AFAIR OCR needle still needs to be anchored to at least one non-OCR region. But later you can get what was read from the screen. Example use at https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/aem.pm#L196-L214

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,98 @@
+    x11_start_program('xterm');
+    send_key('alt-f10');  # maximize xterm to ease troubleshooting
+
+    upload_package_lists;
```
```suggestion
    $self->upload_package_lists;
```

legoktm commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,101 @@
+    assert_script_run('qvm-run -p work -- "rpm -Kv securedrop-workstation-dom0-config-*.rpm"');  # TODO confirm output is correct
+    assert_script_run('qvm-run -p work -- "cat /home/user/securedrop-workstation-dom0-config-*.rpm" > securedrop-workstation.rpm');
+    assert_script_run('sudo dnf -y install securedrop-workstation.rpm');
+    assert_script_run('echo {\"submission_key_fpr\": \"65A1B5FF195B56353CC63DFFCC40EF1228271441\", \"hidserv\": {\"hostname\": \"bnbo6ryxq24fz27chs5fidscyqhw2hlyweelg4nmvq76tpxvofpyn4qd.onion\", \"key\": \"FDF476DUDSB5M27BIGEVIFCFGHQJ46XS3STAP7VG6Z2OWXLHWZPA\"}, \"environment\": \"prod\", \"vmsizes\": {\"sd_app\": 10, \"sd_log\": 5}} | sudo tee /usr/share/securedrop-workstation-dom0-config/config ...
```
If we git cloned the sd-workstation repo (which we'll need to do to run the tests), we'd be able to copy this and the test key.

marmarek replied:
At some point, you probably will want to control what gets installed via some openqa variable. See tests/update2.pm (among other things, it supports adding an extra repo with stuff to test - this is how we run tests on PR content before merging it). But I'd suggest to start with a single simple workflow and add extra flexibility later.

legoktm reviewed (commented):
This is great!

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,101 @@
+    $self->handle_system_startup;
+
+    # Go through launcher
+    assert_and_click("securedrop-launcher", 10);
```
```suggestion
    assert_and_click("securedrop-launcher", timeout => 10);
```

The positional arg form is deprecated and it seems doesn't work in this case.

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,112 @@
+
+    curl_via_netvm;
+
+    assert_script_run('set -eou pipefail'); # Make bash behave like a "normal" language
```
`set -e` will break error reporting (`false; echo $? >/dev/hvc0` no longer works). Other parts are fine, though

marmarek commented:
> I think this is related to the grsec kernel.

FWIW the base test class collects quite a bit of logs, including tarball of the whole /var/log. There, you can find `/var/log/xen/console/guest-sd-small-bookworm-template.log` with this:
```
[2024-10-02 12:21:11] [    1.325369] Kernel panic - not syncing: PAX: this KERNEXEC configuration requires SMEP support
[2024-10-02 12:21:11] [    1.326336] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G                T   6.6.50-1-grsec-workstation #1 d335937bf8be13d0d1eca0b0aca69d85d25a70f1
[2024-10-02 12:21:11] [    1.326336] Call Trace:
[2024-10-02 12:21:11] [    1.326336]  <TASK>
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff81f186ec>] dump_stack_lvl+0x6c/0xe0 ffffffff83c03e60
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff81f18775>] dump_stack+0x15/0x80 ffffffff83c03e80
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff811d1ec1>] panic+0x4e1/0x560 ffffffff83c03e90
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff81bd2f58>] ? intel_init_thermal+0x18/0x3c0 ffffffff83c03ec0
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff8116371a>] identify_cpu+0x8da/0x920 ffffffff83c03f08
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff84c4d1b1>] arch_cpu_finalize_init+0x31/0x2a0 ffffffff83c03f38
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff84c23117>] start_kernel+0x777/0xe00 ffffffff83c03f58
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff84c39b81>] ? x86_64_start_reservations+0x21/0x80 ffffffff83c03fc8
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff84c39b81>] x86_64_start_reservations+0x21/0x80 ffffffff83c03fd0
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff84c39e51>] x86_64_start_kernel+0x151/0x1a0 ffffffff83c03fd8
[2024-10-02 12:21:11] [    1.326336]  [<ffffffff810002d9>] secondary_startup_64_no_verify+0x1b4/0x1db ffffffff83c03fe8
[2024-10-02 12:21:11] [    1.326336]  </TASK>
```

Relevant historical context: https://github.com/QubesOS/qubes-core-admin/commit/75608795b0036708a949328f442cb8150a53f17c, but later (in R4.1 already) reverted: https://github.com/QubesOS/qubes-core-admin/commit/8f38753bdb29ae43c66fb2076142f19168f4423d

And I do see `smep` in /proc/cpuinfo in a HVM on R4.2.

legoktm commented:
@deeplow wrote:
> Sadly I can't trace back to the commit from the test, but I recall this started failing around the time I switched to `yum-qa` instead of `yum`. Do you recall if there is a new grsec kernel in `yum-qa` under testing in the last few days?

Does the timing line up with https://github.com/freedomofpress/securedrop-apt-prod/pull/147 being deployed? (Upgrading from Linux 6.6.31 -> 6.6.50)

Could we dump `/proc/cpuinfo` from a failing run to see if "semp" is really missing or if it's a kernel bug?

I think we could also try to set `pax_softmode=1` in the kernel cmdline to disable some of the protection features, but that's not a long-term fix, because we'd want to test with grsec...

marmarek commented:
Ouch, that indeed looks related. Caviar is Intel Xeon E5-2609 (v1), which seems to not support SMEP yet. Let me add a worker class to exclude this one from your tests...

marmarek commented:
Note to self: SMEP in marketing lingo (ark.intel.com) is called "Intel OS Guard"

marmarek commented:
Here you go: https://openqa.qubes-os.org/tests/114641
Use this one for further restarts.

marmarek commented on tests/securedrop/install.pm:
```diff
@@ -0,0 +1,135 @@
+    # Accept Whonix systemcheck prompts (for sys-whonix and sd-whonix)
+    while (check_screen('whonix-systemcheck-derived-repo-accept', timeout => 300)) {
+        assert_and_click('whonix-systemcheck-derived-repo-accept');
+    }
```
I'd recommend adding `wait_still_screen` in such a loop. Or just `sleep` after clicking. Otherwise it may loop quickly while the window is still visible.

And also, add some timeout (max number of iterations), otherwise if it will keep looping, you'll eventually hit hard timeout and that kills the job and skips collecting logs on failure.

marmarek replied:
You should match just the information you are interested in. If you want to verify specific parts of this message, select just those messages etc. If the test really should fail if that window is a different size, then sure, you can select the whole window, but I doubt that's your goal...

marmarek commented:
> I think this was working before, so something may be up with the server. I'm guessing this is probably a transient issue.

I don't see the curl wrapper loaded in this window. Missing `curl_via_netvm` call?

marmarek commented:
But I see it sometimes work...  Does SD make some firewall changes in sys-net?

marmarek commented:
> _I'm not sure yet if modifying test is OK, since we're going a bit beyond our directory_.

This kind of modification should be okay I think.

marmarek commented:
Test suite name is a lie ;) (it' just says which one I used as a source, later overriding most of its settings).
Anyway, I'd propose the following:
1. Decide on two (or more) variables that will distinguish different jobs (and will cause different tests to be loaded by main.pm). Something like `SECUREDROP_INSTALL=1`, `SECUREDROP_TEST=something`  etc
2. The first one will have also `PUBLISH_HDD_1` set, others will reference it in `HDD_1`.

One thing to decide, how you want to schedule this - should install step be repeated before each test session (meaning tests will run always on "fresh" install)? Or should install be scheduled from time to time (weekly? monthly? something else?) and other tests use the most recent one and start by updating (possibly to a test version)?

marmarek commented:
> Sounds like a fine approach. Going this route, would we be able to express test dependencies [like this](https://openqa.qubes-os.org/tests/116647#dependencies) and run only one "install" job per series of subsequent parallel tests?

Yes, this is the way to go with the "start from fresh install each time" approach. Implementation-wise, it's simple - just name the install job in the START_AFTER_TEST variable in the test jobs (and put them together in the same job group). That last part I think only admin can configure, but I think you can see an example at https://openqa.qubes-os.org/admin/job_templates/8

And FYI, so you get the full picture, those job templates then go through https://openqa.qubes-os.org/admin/test_suites to translate to specific variables. Then, it's up to `main.pm`  to decide what to do - this part you know already.

> Also, would the idea then to have a case-switch (if that's a thing in perl) in `main.pm` directing it to each specific test or would we have a single `.pm` which would receive `SECUREDROP_TEST` as a variable almost like [`system_tests`](https://github.com/QubesOS/openqa-tests-qubesos/blob/7242736b1ae34d9de31e906618ff48055bc853e5/tests/system_tests.pm#L85)? (asking because I don't if both approaches can start multiple tests in parallel).

Whichever is more convenient for you. I guess it mostly depends on how similar those tests are (at the openqa level). "system_tests" file is shared, because from openqa perspective all of those python tests are very similar. But there are others like "gui_tools" one or gui_interactive" that are separate pm files. In any case, there must be some variable(s) which tell which test to run - whether it's routed to separate pm files or the same one doesn't really change much.

> for now we could have it run before each test

Ok

So, I guess the next step is for you to add another test (a variable that should be set on the second job) and possibly also a new pm file. And please document them in README (maybe new section there?). Then I'll configure job group with two jobs (install, test) with those variables set and also having appropriate dependencies.

marmarek commented:
To clarify, the two scenarios are:

## Install once test several times

One job that installs SecureDrop and saves the prepare disk image. This one is running on some schedule (frequency TBD).

The actual tests take that prepared disk image, update it to the version that should be tested and do the tests.  It's up for discussion whether the update happens in each test job separately at the start, or there is yet another intermediate job that takes "old" disk image, update it and saves "new" disk image to run tests on. Initially IMO makes sense to have update in the test job itself, until there will be more than few different test jobs.

BTW When I write "update in the test job", I mean a job in openqa sense. It can (and probably should) be a separate pm file.

Pros: you test also if update from earlier version works.
Cons: it's more complicated (more tasks, jobs).

## Install fresh each time

Here, before each test session, a fresh SecureDrop installation is prepared, so there is no need to do update before each test.

Pros: simpler
Cons: no testing of updates, only fresh install.

---

There is also only slightly related question: do you plan testing things only when they land in some test repo (I guess after merging?). Or also some upcoming changes before they get merged (which in practice probably also will involve putting them in some temporary repo)? It doesn't change much in the architecture, but may influence which option will better fit the use case.

marmarek commented:
Technically, the difference is how I setup job groups, and whether I use `PUBLISH_HDD_1` (reusable disk image, the first variant) or `STORE_HDD_1` (disk image just for the given test run, the second variant).

marmarek commented:
> I was also thinking that the "install fresh each time" option this would make it harder to do test development on the child tests since we'd always have run the parent job, once again

For tests running in VMs (which is your case), no, you don't need to restart parent job. That's relevant only for bare metal tests (where you don't have the convenience of swapping disk image instantly).

marmarek commented:
> The test's overridden name somehow landed inside the test itself

The upload names are built with test name included - see "logs and assets" tab in other jobs, for example https://openqa.qubes-os.org/tests/117619#downloads
It never occurred to me that I could use spaces in test names, and now I know my intuition to avoid spaces was correct :)

marmarek commented:
See https://openqa.qubes-os.org/tests/overview?distri=qubesos&version=4.2&build=2024110618-4.2&groupid=14
You can see the new group on the main page too.

As for extra jobs, technically new pm file doesn't need to be new job - it often makes sense to schedule several tests in a single job (as long as they don't leave system in a weird state even on success). Splitting to separate jobs is mostly an optimization and makes sense if the time gained from parallel execution is significantly greater than overhead of job startup. And depending on test flags (see especially `fatal` and `norollback`), may allow running some tests even though other failed earlier.
So, you can add new tests freely yourself. What you need extra access is splitting them into separate jobs.

For reference, I defined two test suites:
https://openqa.qubes-os.org/admin/test_suites (filter on securedrop)
and then created a job group with them:
https://openqa.qubes-os.org/admin/job_templates/14

marmarek commented:
> I've done 1 now. So `TEST_SECUREDROP` became `SECUREDROP_INSTALL=1` and `SECUREDROP_TEST` to `basic_functionality`. If you could just do the initial setup to set up the dependency / disk situation that would be perfect.

You did in the readme file. But main.pm still uses old variables.

marmarek commented:
You need to restart install job, as the one that completed earlier didn't really install anything due to wrong variable name

marmarek commented:
> One small question: `securedrop_test` is missing `WORKER_CLASS=local,smep`. Should we also have it there? Or because it's dependent, it inherits some settings from the parent job?

It does have "smep" class, no? ("local" is replaced with "hdd_download" because this job downloads HDD image instead of uploading)

marmarek commented:
I still see some spaces in test names. I'm not sure if they upload any logs, but it's asking for troubles...

marmarek commented:
> I was looking in the [test suites](https://openqa.qubes-os.org/admin/test_suites) list:

I see what you mean. The default WORKER_CLASS for SD tests is defined at https://openqa.qubes-os.org/admin/products. "install" test suite overrides it to use "local" class (for faster disk upload).

marmarek commented:
And I think it's time for you to learn how to schedule new test runs, instead of restarting the same one over and over.
1. Fetch API key from web ui (menu in the right top corner)
2. Save it to `.config/openqa/client.conf` in format:
    ```
    [openqa.qubes-os.org]
    key=...
    secret=...
    ```
3. Install openqa-cli (openqa-client package)
4. Schedule new run: `openqa-cli --host=https://openqa.qubes-os.org api isos -X post VERSION=4.2 BUILD=$(date +%Y%m%d%H)-4.2 DISTRI=qubesos ARCH=x86_64 FLAVOR=securedrop CASEDIR="https://github.com/deeplow/openqa-tests-qubesos.git#add-securedrop-test"`
 
Some explanation for the parameters: generally, parameters to the command are variables to be set on all jobs. `VERSION`, `BUILD`, `ARCH`, `DISTRI` and `FLAVOR` are mandatory. `BUILD` should be something unique for this run - as you can see, I use date-based one, but it can be anything (probably avoid spaces). Use `ARCH`, `DISTRI` and `FLAVOR` as shown above. With `VERSION` you can choose which Qubes version (currently 4.2 and 4.3 should work) - for this to work, I guess your tests would need to work on given version (for example 4.3 will want packages compatible with F41 dom0). The `CASEDIR` makes it use your branch.

You can find more example in README how we use those global variables in other tests (things like choosing what templates to update or test etc).
There is no way (that I know of) to set variable on just one job at this stage. But if you need you can have variables that for example only one "pm" file use.

---
So, I guess once the current version is green, we can merge this PR, and you can develop further changes in a new PR.

marmarek commented:
Watch video closely, there was some window that got dismissed by this very space.

marmarek commented:
> One thing that could triggering this is that I'm running `curl_via_netvm` twice.

It shouldn't harm, it just defines `curl` as a shell function.

Is there some sys-net change? Idk, some firewall, or maybe some weird routing everything via tor? Or maybe it fails on reading the input file for some reason? Maybe some shell startup script (.bashrc etc) that could swallow/redirect all input?

My next idea is to debug it via developer mode, let me try

marmarek commented:
> After entering that mode, one needs to use VNC, correct?

In this case, yes (which isn't exposed to the internet directly). The other use case for developer mode is to create needles interactively.

marmarek commented:
Ok, this is really weird, it looks like some network packets are lost... My current hypothesis is it doesn't like sys-net restart (kernel or qemu bug). I'll try to get smaller reproducer.
In the meantime, can you try to correlate this issue with worker the test runs on?

marmarek commented:
I just tried to reproduce in a fresh job before SD gets installed, and it always works, also after sys-net restart, qubes restart and qubes+sys-net restart...

marmarek commented:
Time is unlikely the factor here, we have jobs running for 3h+ that upload a lot of logs (including multi-MB tarball of the whole /var/log) at the end.

marmarek commented:
So, I have no idea what is going on... it's probably some QEMU or e1000e driver issue, but I don't really know why it happens, and why only here.
I kinda suspect sys-net restart is related in some way, but not sure how exactly.
If I restart sys-net at the start of the job, I can get similar failure eventually, but very rarely (maybe 1 in 100 attempts or even less). And then reloading `e1000e` module in sys-net seems to help. I tried restarting sys-net few more times and results are inconclusive - at some times it made it better, but at other times it made it worse. Maybe it isn't just about restarting sys-net but some other thing at the same time?
But at the end it fails most of the time (idk, 99% of cases or such), and reloading the module doesn't help. Restarting sys-net and reloading the module doesn't help either.

All of that I tested by doing curl call in sys-net directly, to eliminate few moving parts.

I'll test some workarounds like adding `--retry` and `--max-time` to the curl call...

<details><summary>Details</summary>
<p>

This is how it looks on tcpdump from the host side (packets coming from qemu appear as coming from localhost):
```
04:36:44.072345 IP (tos 0x0, ttl 64, id 16376, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.36244 > 127.0.0.1.20043: Flags [S], cksum 0xfe30 (incorrect -> 0xf201), seq 87966507, win 65495, options [mss 65495,sackOK,TS val 3799537470 ecr 0,nop,wscale 7], length 0
04:36:44.072375 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.20043 > 127.0.0.1.36244: Flags [S.], cksum 0xfe30 (incorrect -> 0x4a01), seq 2951528023, ack 87966508, win 65483, options [mss 65495,sackOK,TS val 3799537470 ecr 3799537470,nop,wscale 7], length 0
04:36:44.072399 IP (tos 0x0, ttl 64, id 16377, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.36244 > 127.0.0.1.20043: Flags [.], cksum 0xfe28 (incorrect -> 0x70bd), ack 1, win 512, options [nop,nop,TS val 3799537470 ecr 3799537470], length 0
04:36:44.074104 IP (tos 0x0, ttl 64, id 16378, offset 0, flags [DF], proto TCP (6), length 280)
    127.0.0.1.36244 > 127.0.0.1.20043: Flags [P.], cksum 0xff0c (incorrect -> 0xa859), seq 1:229, ack 1, win 512, options [nop,nop,TS val 3799537471 ecr 3799537470], length 228
04:36:44.074125 IP (tos 0x0, ttl 64, id 22166, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20043 > 127.0.0.1.36244: Flags [.], cksum 0xfe28 (incorrect -> 0x6fd9), ack 229, win 510, options [nop,nop,TS val 3799537471 ecr 3799537471], length 0
04:36:44.281074 IP (tos 0x0, ttl 64, id 16379, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.36244 > 127.0.0.1.20043: Flags [P.], cksum 0x03dd (incorrect -> 0xc4bc), seq 229:1689, ack 1, win 512, options [nop,nop,TS val 3799537678 ecr 3799537471], length 1460
04:36:44.281098 IP (tos 0x0, ttl 64, id 22167, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20043 > 127.0.0.1.36244: Flags [.], cksum 0xfe28 (incorrect -> 0x6818), ack 1689, win 621, options [nop,nop,TS val 3799537678 ecr 3799537678], length 0
04:36:44.283811 IP (tos 0x0, ttl 64, id 16380, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.36244 > 127.0.0.1.20043: Flags [P.], cksum 0x03dd (incorrect -> 0xaca3), seq 1689:3149, ack 1, win 512, options [nop,nop,TS val 3799537681 ecr 3799537678], length 1460
04:36:44.283833 IP (tos 0x0, ttl 64, id 22168, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20043 > 127.0.0.1.36244: Flags [.], cksum 0xfe28 (incorrect -> 0x625e), ack 3149, win 621, options [nop,nop,TS val 3799537681 ecr 3799537681], length 0
```
and no further packets for this connection (not even retransmissions), looks like qemu isn't sending them. When looking from the sys-net side, all are sent, and there are retransmissions for the missing ones.

when it works, it looks like this:
```
06:45:54.364797 IP (tos 0x0, ttl 64, id 65392, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [S], cksum 0xfe30 (incorrect -> 0xae12), seq 2227083767, win 65495, options [mss 65495,sackOK,TS val 3807287762 ecr 0,nop,wscale 7], length 0
06:45:54.364817 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [S.], cksum 0xfe30 (incorrect -> 0xd332), seq 2499070244, ack 2227083768, win 65483, options [mss 65495,sackOK,TS val 3807287762 ecr 3807287762,nop,wscale 7], length 0
06:45:54.364833 IP (tos 0x0, ttl 64, id 65393, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [.], cksum 0xfe28 (incorrect -> 0xf9ee), ack 1, win 512, options [nop,nop,TS val 3807287762 ecr 3807287762], length 0
06:45:54.366026 IP (tos 0x0, ttl 64, id 65394, offset 0, flags [DF], proto TCP (6), length 300)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0xff20 (incorrect -> 0xd07c), seq 1:249, ack 1, win 512, options [nop,nop,TS val 3807287763 ecr 3807287762], length 248
06:45:54.366040 IP (tos 0x0, ttl 64, id 11885, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xf8f6), ack 249, win 510, options [nop,nop,TS val 3807287763 ecr 3807287763], length 0
06:45:54.366237 IP (tos 0x0, ttl 64, id 65395, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x377b), seq 249:1709, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287763], length 1460
06:45:54.366247 IP (tos 0x0, ttl 64, id 11886, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xf2d1), ack 1709, win 621, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366277 IP (tos 0x0, ttl 64, id 65396, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x0746), seq 1709:3169, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366283 IP (tos 0x0, ttl 64, id 11887, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xed24), ack 3169, win 614, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366300 IP (tos 0x0, ttl 64, id 65397, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xe15d), seq 3169:4629, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366311 IP (tos 0x0, ttl 64, id 11888, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xe777), ack 4629, win 607, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366325 IP (tos 0x0, ttl 64, id 65398, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x61fa), seq 4629:6089, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366330 IP (tos 0x0, ttl 64, id 11889, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xe1ca), ack 6089, win 600, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366344 IP (tos 0x0, ttl 64, id 65399, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x3073), seq 6089:7549, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366349 IP (tos 0x0, ttl 64, id 11890, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xdc1d), ack 7549, win 593, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366491 IP (tos 0x0, ttl 64, id 65400, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xf323), seq 7549:9009, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366498 IP (tos 0x0, ttl 64, id 11891, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xd670), ack 9009, win 586, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366518 IP (tos 0x0, ttl 64, id 65401, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xf3b9), seq 9009:10469, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366523 IP (tos 0x0, ttl 64, id 11892, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xd0c3), ack 10469, win 579, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366537 IP (tos 0x0, ttl 64, id 65402, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x0adf), seq 10469:11929, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366542 IP (tos 0x0, ttl 64, id 11893, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xcb16), ack 11929, win 572, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366556 IP (tos 0x0, ttl 64, id 65403, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x3c53), seq 11929:13389, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366561 IP (tos 0x0, ttl 64, id 11894, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xc569), ack 13389, win 565, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366574 IP (tos 0x0, ttl 64, id 65404, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xdefd), seq 13389:14849, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366579 IP (tos 0x0, ttl 64, id 11895, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xbfbc), ack 14849, win 558, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366686 IP (tos 0x0, ttl 64, id 65405, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xbc08), seq 14849:16309, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366706 IP (tos 0x0, ttl 64, id 65406, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0x65c1), seq 16309:17769, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366713 IP (tos 0x0, ttl 64, id 11896, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xb9c9), ack 16309, win 621, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366716 IP (tos 0x0, ttl 64, id 11897, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xb420), ack 17769, win 610, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366733 IP (tos 0x0, ttl 64, id 65407, offset 0, flags [DF], proto TCP (6), length 1512)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x03dd (incorrect -> 0xd87e), seq 17769:19229, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1460
06:45:54.366738 IP (tos 0x0, ttl 64, id 11898, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xae73), ack 19229, win 603, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.366753 IP (tos 0x0, ttl 64, id 65408, offset 0, flags [DF], proto TCP (6), length 1280)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [P.], cksum 0x02f5 (incorrect -> 0xbfed), seq 19229:20457, ack 1, win 512, options [nop,nop,TS val 3807287764 ecr 3807287764], length 1228
06:45:54.366758 IP (tos 0x0, ttl 64, id 11899, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [.], cksum 0xfe28 (incorrect -> 0xa9ad), ack 20457, win 597, options [nop,nop,TS val 3807287764 ecr 3807287764], length 0
06:45:54.369007 IP (tos 0x0, ttl 64, id 11900, offset 0, flags [DF], proto TCP (6), length 241)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [P.], cksum 0xfee5 (incorrect -> 0x32ca), seq 1:190, ack 20457, win 621, options [nop,nop,TS val 3807287766 ecr 3807287764], length 189
06:45:54.369021 IP (tos 0x0, ttl 64, id 65409, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [.], cksum 0xfe28 (incorrect -> 0xa942), ack 190, win 511, options [nop,nop,TS val 3807287766 ecr 3807287766], length 0
06:45:54.372973 IP (tos 0x0, ttl 64, id 65410, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [F.], cksum 0xfe28 (incorrect -> 0xa93c), seq 20457, ack 190, win 512, options [nop,nop,TS val 3807287770 ecr 3807287766], length 0
06:45:54.373165 IP (tos 0x0, ttl 64, id 11901, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.20023 > 127.0.0.1.47168: Flags [F.], cksum 0xfe28 (incorrect -> 0xa8ca), seq 190, ack 20458, win 621, options [nop,nop,TS val 3807287770 ecr 3807287770], length 0
06:45:54.373182 IP (tos 0x0, ttl 64, id 65411, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.47168 > 127.0.0.1.20023: Flags [.], cksum 0xfe28 (incorrect -> 0xa937), ack 191, win 512, options [nop,nop,TS val 3807287770 ecr 3807287770], length 0
```

</p>
</details>

marmarek commented:
as for a temporary band-aid, try to add something like this to your branch:
```patch
diff --git a/lib/networking.pm b/lib/networking.pm
index c0053f2..c1d4215 100644
--- a/lib/networking.pm
+++ b/lib/networking.pm
@@ -102,7 +102,7 @@ curl() {
     fi
 
     if [ -n "\$inputfile" ]; then
-        qvm-run --no-gui -p sys-net "curl \$allargs" <\$inputfile
+        qvm-run --no-gui -p sys-net "curl --max-time 10 \$allargs" <\$inputfile
     else
         qvm-run --no-gui -p sys-net "curl \$allargs"
     fi
```
and `failok=>1`

It won't fix uploading, but at least won't block the whole thing on this issue. And when we'll manage to fix uploading, it should automatically start working again without test change.

marmarek commented on tests/securedrop/upload_packages.pm:
```diff
@@ -0,0 +1,59 @@
+
+    curl_via_netvm;
+
+    my $fname = $self->save_and_upload_log('rpm -qa', 'dom0-packages.txt');
```
this one is missing failok flag

marmarek commented on tests/securedrop/upload_packages.pm:
```diff
@@ -0,0 +1,59 @@
+        next unless /Template/;
+        s/\|.*//;
+        $fname = $self->save_and_upload_log("qvm-run --no-gui -ap $_ 'rpm -qa; dpkg -l; pacman -Q; true'",
+                "template-$_-packages.txt", ('timeout' => 90, 'failok' => 1));
```
I'm not an perl expert by any means, but in other places I see a syntax like this:
```suggestion
                "template-$_-packages.txt", timeout => 90, failok => 1);
```
So, without extra quotes or parenthesis.

marmarek commented on lib/installedtest.pm:
```diff
@@ -288,7 +288,12 @@ sub maybe_unlock_screen {
 sub save_and_upload_log {
     my ($self, $cmd, $file, $args) = @_;
```
And now I wonder if this shouldn't be
```suggestion
    my ($self, $cmd, $file, %args) = @_;
```
That's what I see in testapi.pm for example.

marmarek commented:
> # Test died: Can't open file "ulogs/upload_packages-dom0-packages.txt": No such file or directory at /var/lib/openqa/pool/4/openqa-tests-qubesos/tests/securedrop/upload_packages.pm line 32.

Upload failed after all...

marmarek committed:
Move uploading packages list to a base class

Make it reusable, deduplicate the code (it's already in two places).
While at it, add support for failok argument and log info if it actually failed
to upload.

marmarek committed:
Adjust sut_packages.txt format to prefix each line with VM name

Diff at openqa webui doesn't show context at all. While it's possible to guess
from where a given package is based in the version format, lets make it easier
by making it explicit.

marmarek commented:
Ok, I think I got it working: https://github.com/QubesOS/openqa-tests-qubesos/pull/27, feel free to merge/rebase/whatever

marmarek commented:
FYI now that the install job worked, you can click restart on just the second job and not wait for re-install (assuming that install was okay and you don't want to change something at that stage).

marmarek commented:
The failing assert_screen x11 is part of switching back to the graphical console - it expects no windows open... If SD client autostart is expected, then I suggest making a needle with `x11` tag (and maybe also some "securedrop-login" or such, even if unused for now?).

marmarek commented:
> then I rebase this PR on top of main

you already did ;)

As for merging this PR, I'd prefer it to be green (even if the second job - the actual test - is more or less empty). For example I see the SECUREDROP_TEST variable was set to wrong value (I fixed it in settings now). If I'm correct about the X11 needle situation, it's just a matter of adding it via needle editor.

marmarek commented:
@deeplow unrelated to this PR, but still about openqa tests:
https://openqa.qubes-os.org/tests/125341#downloads `guest-sd-large-bookworm-template.log`:
```
[2025-01-09 09:19:39] [    0.796063] Run /init as init process
[2025-01-09 09:19:39] Loading, please wait...
[2025-01-09 09:19:39] Starting systemd-udevd version 252.31-1~deb12u1

[2025-01-09 09:19:39] [    1.075869] Invalid max_queues (4), will use default max: 2.
[2025-01-09 09:19:40] [    1.157320] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled; bounce buffer: enabled
[2025-01-09 09:19:40] [    1.180986]  xvda: xvda1 xvda2 xvda3
[2025-01-09 09:19:40] [    1.195462] blkfront: xvdb: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled; bounce buffer: enabled
[2025-01-09 09:19:40] [    1.211144] blkfront: xvdc: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled; bounce buffer: enabled
[2025-01-09 09:19:40] Begin: Loading essential drivers ... [    1.497039] device-mapper: core: CONFIG_IMA_DISABLE_HTABLE is disabled. Duplicate IMA measurements will not be recorded in the IMA log.
[2025-01-09 09:19:40] [    1.497183] device-mapper: uevent: version 1.0.3
[2025-01-09 09:19:40] [    1.497935] device-mapper: ioctl: 4.48.0-ioctl (2023-03-01) initialised: dm-devel@redhat.com
[2025-01-09 09:19:40] done.
[2025-01-09 09:19:40] Begin: Running /scripts/init-premount ... done.
[2025-01-09 09:19:40] Begin: Mounting root file system ... Begin: Running /scripts/local-top ... Begin: Waiting for /dev/xvda* devices... ... [2025-01-09 09:19:40] done.
[2025-01-09 09:19:40] Begin: Qubes: Doing R/W setup for TemplateVM... ... [2025-01-09 09:19:40] [    1.903477]  xvdc: xvdc1 xvdc3
[2025-01-09 09:19:40] Setting up swapspace version 1, size = 1073737728 bytes
[2025-01-09 09:19:40] UUID=038db4ec-7460-494c-91b9-1eb34836714b
[2025-01-09 09:19:40] [    1.973225] Adding 1048572k swap on /dev/xvdc1.  Priority:-2 extents:1 across:1048572k SS
[2025-01-09 09:19:41] done.
[2025-01-09 09:19:41] done.
[2025-01-09 09:19:41] Begin: Running /scripts/local-premount ... done.
[2025-01-09 09:19:41] Begin: Waiting for root file system ... [2025-01-09 09:19:42] Begin: Running /scripts/local-block ... done.
[2025-01-09 09:19:43] Begin: Running /scripts/local-block ... done.
[2025-01-09 09:19:44] Begin: Running /scripts/local-block ... done.
[2025-01-09 09:19:45] Begin: Running /scripts/local-block ... done.
[2025-01-09 09:19:46] Begin: Running /scripts/local-block ... done.
[2025-01-09 09:19:47] Begin: Running /scripts/local-block ... done.
...
[2025-01-09 09:20:11] Gave up waiting for root file system device.  Common problems:
```

Is it something you know about, or maybe need help with? This part is normally quite reliable...

marmarek commented:
FYI the fedora-41 update issue you've hit recently is https://github.com/QubesOS/qubes-issues/issues/9744, it's fixed now.

marmarek commented:
In one of the logs I see this:
```
2025-02-27 17:10:49,927 - detail.sdw_updater.Updater:185(_qubes_updater_parse_stdout) INFO: [Qubes updater] fedora-41-xfce:out: ERROR (exception [Errno 13] Permission denied: '/var/log/qubes/update-fedora-41-xfce.log')
```
Looks like calling qubes-vm-update sometimes as root and sometimes as user. Is it something on the SD side, or somewhere else? I ask, because I know qubes tests do call qubes-vm-update sometimes as root, but I don't think you re-use this part. But if that's something you find in some base code, or maybe pre-existing in the disk image, let me know.

marmarek commented:
What kind of issues? I know the session expires after some time, but it's usually enough to just click "login" again.

marmarek commented:
> Btw, I've been having some GitHub authentication issues on OpenQA

https://github.com/os-autoinst/openQA/pull/6236 ... anyway, should work now.

marmarek commented:
I guess you are missing sending it to the VM: https://github.com/QubesOS/openqa-tests-qubesos/blob/main/tests/update2.pm#L37-L42

But also, it should be there in the base image already (it's done in the "install_fixups" stage), just not the version with your modifications.

marmarek commented:
> Thanks! After exploring this route I ended up running into strange error cases. The test would have no assets uploaded ([not even video](https://openqa.qubes-os.org/tests/139694#))

I've seen those, and I'm also confused. Looking at the message, it looks like some result property is NULL. You experimented with custom test module names, maybe somewhere you used an empty one or something like this? Or a space somewhere?

marmarek commented:
BTW, I very much welcome your convert_junit.py change, in context of https://github.com/QubesOS/qubes-issues/issues/9898

marmarek commented:
Check now (without changes on your side)
I disabled auto-update, I think default for this option changed recently.

marmarek commented:
@deeplow sorry, I might have broken SD tests by enabling presentation mode in the base image already...

marmarek commented:
There are two options:
1. standard base image job (part of weekly builds) -> SD preparation job (creates "SD base image") - the latter would run either after weekly builds, or on some other schedule
2. separate base image job, with SD-specific preparation included already - running in parallel to other weekly jobs

The second one may not be very time-effective, as it's doing installation again, but it's done on a schedule (doesn't delay PR results), and have some benefits: you can adjust installation itself - for example don't install templates you don't need.

BTW, part of what you need is already implemented - for example you can set `INSTALL_TEMPLATES=debian-12-minimal` to install extra templates. I try to document supported variables in README, but I'm sure some are missing there...
In fact, there is already a job that adds debian-12-minimal template too (and few others), see https://openqa.qubes-os.org/tests/151215

Both approaches would need some variable (can be `SECUREDROP_PREP`) and relevant main.pm snippet in the same place. The difference is only which other variables will be set in that job (especially - if ISO=... for fresh install, or HDD_1 for extending existing base image).

marmarek commented:
@deeplow [recent failure](https://openqa.qubes-os.org/tests/154927) has this in xen log:
```
(XEN) d14v0 Triple fault - invoking HVM shutdown action 1
(XEN) *** Dumping Dom14 vcpu#0 state: ***
(XEN) ----[ Xen-4.17.5  x86_64  debug=n  Not tainted ]----
(XEN) CPU:    0
(XEN) RIP:    0008:[<00000000000f5eee>]
(XEN) RFLAGS: 0000000000010006   CONTEXT: hvm guest (d14v0)
(XEN) rax: 0000000000000060   rbx: 000000002030e623   rcx: 0000000085fffffe
(XEN) rdx: 00000000188e7fd6   rsi: 0000000000000000   rdi: 00000000008dc878
(XEN) rbp: 0000000000000000   rsp: 00000000188e7fec   r8:  0000000000000000
(XEN) r9:  0000000000000000   r10: 0000000000000000   r11: 0000000000000000
(XEN) r12: 0000000000000000   r13: 0000000000000000   r14: 0000000000000000
(XEN) r15: 0000000000000000   cr0: 0000000000000011   cr4: 0000000000000000
(XEN) cr3: 0000000000000000   cr2: 0000000000000000
(XEN) fsb: 0000000000000000   gsb: 0000000000000000   gss: 0000000000000000
(XEN) ds: 0010   es: 0010   fs: 0010   gs: 0010   ss: 0010   cs: 0008
```
But IIUC that test was from the main branch of securedrop-workstation repo, not a PR. I don't think there were any recent changes in Xen, so maybe some change in the VM kernel? Which one is used there (for sys-net)? Is it maybe a grsec one (I think it isn't but just to be sure)? Recent test run of R4.2 updates (no SD) with default kernel (6.12.47) was okay...

marmarek commented:
but there is
```
[2025-10-06 14:15:01] (XEN) [VT-D]DMAR:[DMA Write] Request device [0000:00:03.0] fault addr fe1457b4fe145000
[2025-10-06 14:15:01] (XEN) [VT-D]DMAR: reason 04 - Access beyond MGAW
```
and the other one also has those, but a bit earlier

What is most interesting here, is that sys-net (when it starts correctly) doesn't actually have 0000:00:03.0 device at all...

deeplow commented:
Perfect! Thanks!

deeplow commented:
I should have looked at the real repo. Thanks!

deeplow commented:
FYI, I've noticed some failures that don't seem to be SecureDrop-related:
- https://openqa.qubes-os.org/tests/164739
- https://openqa.qubes-os.org/tests/164830

It doesn't always happen, but I think  for the time being we can work around it. And it doesn't seem to be OpenQA worker-related, this time.

marmarek commented:
Ramen has some slower download speed for some reason, you can hit restart and it should be good. The actual download continues in the background, just test stopped waiting for it.

marmarek commented:
> * **added scenario**:
>       
>       * `securedrop_install` → `securedrop_upgrade` → other `securedrop_*` tests

I set it this way now. The middle job got `SECUREDROP_UPGRADE=1`. If you don't have a need to do any upgrade, you can control what the middle job is doing using other variables (basically making it a no-op if needed).

AFAIR a while ago we also discussed having the install job running on schedule, and other tests using disk saved by it (and not repeat install every time) - is it something you'd be still interested in?

deeplow commented:
> The middle job got SECUREDROP_UPGRADE=1. If you don't have a need to do any upgrade, you can control what the middle job is doing using other variables (basically making it a no-op if needed).

That's a good idea and this way we only really need one scenario. We may even get away with overriding that variable at the entire job creation. The default performs an upgrade, overriding to disable it, does a no-op.

> AFAIR a while ago we also discussed having the install job running on schedule, and other tests using disk saved by it (and not repeat install every time) - is it something you'd be still interested in?

Yes. That would still be relevant for us. That would actually be two disk images: one that just has the "pre-caching" job (and on which a clean install can be performed) and one which has the pre-caching and the production installation.

If I understand correctly the only trick to make this work is to have a `PUBLISH_HDD_1=` set. And whenever that job runs again it just writes over the previous artifact, meaning that any test using it a it's `HDD_1` will simply use the latest one. If so, do you mind if I play a bit with this the HDD uploading myself? (just making sure I don't break anything or leave dangling data that doesn't get rotated out).

marmarek commented:
> If I understand correctly the only trick to make this work is to have a `PUBLISH_HDD_1=` set. And whenever that job runs again it just writes over the previous artifact, meaning that any test using it a it's `HDD_1` will simply use the latest one.

Yes

> If so, do you mind if I play a bit with this the HDD uploading myself? (just making sure I don't break anything or leave dangling data that doesn't get rotated out).

+1
